src/main/java/bio/terra/common/exception/SamApiException.java

```diff
@@ -0,0 +1,86 @@
+package bio.terra.common.exception;
+
+import bio.terra.common.retry.SamRetry;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.api.client.http.HttpStatusCodes;
+import java.util.Collections;
+import java.util.List;
+import org.apache.commons.lang3.StringUtils;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.broadinstitute.dsde.workbench.client.sam.model.ErrorReport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * Wrapper exception for non-200 responses from calls to Sam. Does not extend ErrorReportException
+ * so that SamApiException is a checked exception type. Can be converted to an ErrorReportException
+ * subtype using the toErrorReportException function.
+ * */
+public class SamApiException extends Exception {
+  private ApiException apiException;
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+  private static Logger logger = LoggerFactory.getLogger(SamApiException.class);
+
+  static public SamApiException createSamApiException(ApiException apiException) {
+    // Sometimes the sam message is buried one level down inside of the error report object.
+    // If we find an empty message then we try to deserialize the error report and use that message.
+    String message = apiException.getMessage();
+    if (StringUtils.isEmpty(message)) {
+      try {
+        ErrorReport errorReport =
+            objectMapper.readValue(apiException.getResponseBody(), ErrorReport.class);
+        message = errorReport.getMessage();
+      } catch (JsonProcessingException ex) {
+        logger.debug("Unable to deserialize sam exception response body");
+      }
+    }
+    return new SamApiException(message, apiException);
+  }
+
+  public SamApiException(String message, ApiException apiException) {
```

Comment:
Do we actually want this to be public, or should we always use the factory function?

Reply:
Good point.